src/store/useRoslib.ts

```diff
@@ -0,0 +1,152 @@
+import ROSLIB from 'roslib';
+import { ref } from 'vue';
+import type { Ref } from 'vue';
+
+const HEARTBEAT_DISCONNECT_SECONDS = 2;
+const RECONNECTION_GRACE_SECONDS = 2;
+const SECONDS_TO_TIMESTAMP = 1000;
+
+export const useRoslibStore = defineStore('roslib', () => {
+  const ros = new ROSLIB.Ros({ url: undefined });
+  const isWebSocketConnected = ref<boolean>(false);
+  const stop = ref<boolean>(false);
+  function init(serverHost: string) {
+    ros.connect(serverHost);
+    console.count('Reinitialized RosInit');
+    // TODO: Realistically, we should use websocket ping/pong control frames for this.
+    //       This potentially requires hooking into rosbridge_server, so we'll leave it for the future.
+    //       With this, we can increase the heartbeat frequency.
+    let heartbeatTime: number = Date.now() + RECONNECTION_GRACE_SECONDS * SECONDS_TO_TIMESTAMP;
+
+    ros.on('connection', () => {
+      isWebSocketConnected.value = true;
+    });
+    ros.on('error', (error) => {
+      isWebSocketConnected.value = false;
+      console.error(error);
+    });
+    ros.on('close', () => {
+      isWebSocketConnected.value = false;
+    });
+
+    const heartbeatRes = new ROSLIB.Topic({
+      ros,
+      name: '/hbr',
+      messageType: 'std_msgs/Bool',
+      compression: 'cbor',
+    });
+
+    heartbeatRes.subscribe<boolean>((msg) => {
+      // TODO: Why are we using bool for a heartbeat?
+      if (msg.data) {
+        heartbeatTime = Date.now();
+      }
+    });
+
+    // Close the connection if the heartbeat stops for too long.
+    const interval = setInterval(() => {
+      if (stop.value) {
+        clearInterval(interval);
+
+        // No need to error if the websocket is already closed.
+        try {
+          ros.close();
+        } catch (_) {
+          /* empty */
+        }
+
+        return;
+      }
+      if (Date.now() - heartbeatTime > HEARTBEAT_DISCONNECT_SECONDS * SECONDS_TO_TIMESTAMP) {
+        isWebSocketConnected.value = false;
+
+        // No need to error if the websocket is already closed.
+        try {
+          ros.close();
+        } catch (_) {
+          /* empty */
+        }
+
+        // Forcefully reconnect.
+        // @ts-expect-error socket is private but needs to be set to null to force roslib to reconnect in case of hangs.
+        ros.socket = null;
+        // Give the reconnection extra time so that it doesn't
+        // immediately get killed.
+        heartbeatTime = Date.now() + RECONNECTION_GRACE_SECONDS * SECONDS_TO_TIMESTAMP;
+        ros.connect(serverHost);
+      }
+    }, 100);
+  }
+  /**
+   * Generic Subscriber to interact with Ros
+   * @param topicName should start with '/' along with topic name
+   * @param messageType Ros Message Type
+   * @param defaultValue? optional starting value
+   * @returns Vue Ref that can be undefined if not provided default value. Can use "as Ref<T>" if default Value is provided
+   */
+  function subscribe<T>(
```

Comment:
sub and pub (and anything else that needs to make ros objects) should probably be cached rather than creating a new one every time, especially for frequent publishes.

Reply:
for reference, here is the topic source code: https://github.com/RobotWebTools/roslibjs/blob/develop/src/core/Topic.js

you can see that it is heavily oop based, so the object really should be created and used multiple times. This is unlike the `Message` class that can be thrown away once created

Reply:
Thanks for linking to the source code. This actually helps a separate problem I had with unsubscribe as from my understanding it needs to be the same Topic object to be able to unsub. I'll experiment with it when I implement unsubscribe, but I am not 100% sure this would actually impact performance.

TLDR: each sub needs a topic object to be able to unsubscribe due to how they coded it up. For Pubs, I will look into if caching helps when I implement unsub